fish-rust/src/builtins/random.rs

```diff
@@ -0,0 +1,193 @@
+    STATUS_CMD_OK, STATUS_INVALID_ARGS,
+};
+use crate::ffi::{parser_t};
+use crate::wchar::{widestrs, wstr};
+use crate::wgetopt::{wgetopter_t, wopt, woption, woption_argument_t};
+use crate::wutil::{self, format::printf::sprintf, fish_wcstoi_radix_all, wgettext_fmt};
+use num_traits::PrimInt;
+use rand::{Rng, SeedableRng};
+use rand::rngs::SmallRng;
+use once_cell::sync::Lazy;
+use std::sync::Mutex;
+
+static seeded_engine: Lazy<Mutex<SmallRng>> = Lazy::new(|| { Mutex::new(SmallRng::from_entropy()) });
+
+
+#[widestrs]
+pub fn random(
+    parser: &mut parser_t,
+    streams: &mut io_streams_t,
+    argv: &mut [&wstr],
+) -> Option<c_int> {
+    let cmd = argv[0];
+    let argc = argv.len();
+    let print_hints = false;
+
+    const shortopts: &wstr = "+:h"L;
+    const longopts: &[woption] = &[
+        wopt("help"L, woption_argument_t::no_argument, 'h'),
+    ];
+
+    let mut w = wgetopter_t::new(shortopts, longopts, argv);
+    while let Some(c) = w.wgetopt_long() {
+        match c {
+            'h' => {
+                builtin_print_help(parser, streams, cmd);
+                return STATUS_CMD_OK;
+            }
+            ':' => {
+                builtin_missing_argument(parser, streams, cmd, argv[w.woptind - 1], print_hints);
+                return STATUS_INVALID_ARGS;
+            }
+            '?' => {
+                builtin_unknown_option(parser, streams, cmd, argv[w.woptind - 1], print_hints);
+                return STATUS_INVALID_ARGS;
+            }
+            _ => {
+                panic!("unexpected retval from wgeopter.next()");
+            }
+        }
+    }
+
+    let mut engine = seeded_engine.lock().unwrap();
+    let mut start = 0;
+    let mut end = 32767;
+    let mut step = 1;
+    let arg_count = argc - w.woptind;
+    let i = w.woptind;
+    if arg_count >= 1 && argv[i] == "choice" {
+        if arg_count == 1 {
+            streams.err.append(wgettext_fmt!(
+                "%ls: nothing to choose from\n",
+                cmd,
+            ));
+            return STATUS_INVALID_ARGS;
+        }
+
+        let rand = engine.gen_range(0..arg_count - 1);
+        streams.out.append(sprintf!("%ls\n"L, argv[i + 1 + rand]));
+        return STATUS_CMD_OK;
+    }
+    fn parse<T: PrimInt>(streams: &mut io_streams_t, cmd: &wstr, num : &wstr) -> Result<T, wutil::Error> {
+        let res = fish_wcstoi_radix_all(num.chars(), None, true);
+        if res.is_err() {
+            streams.err.append(wgettext_fmt!(
+                "%ls: %ls: invalid integer\n",
+                cmd,
+                num,
+            ));
+        }
+        return res;
+    }
+
+    match arg_count {
+        0 => {
+            // Keep the defaults
+        },
+        1 => {
+            // Seed the engine persistently
+            let num = parse::<i64>(streams, cmd, argv[i]);
+            match num {
+                Err(_) => return STATUS_INVALID_ARGS,
+                Ok(x) => *engine = SmallRng::seed_from_u64(x as u64)
```

Comment:
In general `as` casts are iffy and should be avoided for anything other than float<->int conversion, but I guess here it's what we want (converting negative values to large positive values).

Reply:
Yeah, this just wants "the bits".